src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.event;
+
+import com.mojang.datafixers.util.Pair;
+import java.util.Map;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.world.level.Level;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
+import net.neoforged.fml.LogicalSide;
+import net.neoforged.fml.event.IModBusEvent;
+import net.neoforged.neoforge.client.DimensionTransitionScreenManager.ReceivingLevelScreenFactory;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * <p>Event for registering screen effects when transitioning across dimensions.
+ * Note that there is a priority order when it comes to what screens are displayed: <br>
+ * - Using registerConditionalEffect has priority over the usual transition effects, and will only fire when travelling to the specified dimension coming from a certain dimension. <br>
+ * - If a dimension has a screen that displays when entering it, that will have priority over a dimension that has one when you leave it. </p>
+ *
+ * <p>This event is not {@linkplain ICancellableEvent cancellable}, and does not {@linkplain HasResult have a result}.
+ *
+ * <p>This event is fired on the mod-specific event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterDimensionTransitionScreenEvent extends Event implements IModBusEvent {
+    private final Map<Pair<ResourceKey<Level>, ResourceKey<Level>>, ReceivingLevelScreenFactory> conditionalDimensionEffects;
+    private final Map<ResourceKey<Level>, ReceivingLevelScreenFactory> toEffects;
+    private final Map<ResourceKey<Level>, ReceivingLevelScreenFactory> fromEffects;
+
+    @ApiStatus.Internal
+    public RegisterDimensionTransitionScreenEvent(Map<Pair<ResourceKey<Level>, ResourceKey<Level>>, ReceivingLevelScreenFactory> conditionalDimensionEffects,
+            Map<ResourceKey<Level>, ReceivingLevelScreenFactory> toEffects,
+            Map<ResourceKey<Level>, ReceivingLevelScreenFactory> fromEffects) {
+        this.conditionalDimensionEffects = conditionalDimensionEffects;
+        this.toEffects = toEffects;
+        this.fromEffects = fromEffects;
+    }
+
+    /**
+     * Registers a dimension transition when traveling to a dimension.
+     */
+    public void registerIncomingEffect(ResourceKey<Level> dimension, ReceivingLevelScreenFactory screen) {
+        this.toEffects.put(dimension, screen);
```

Comment:
I would suggest using putIfAbsent() and returning the result. Later mods silently overwriting earlier mods leads to hard to debug errors. With putIfAbsent(), the mod that gets its registration denied will at least get notified about it.

**_If_** there is a need to allow later mods to overwrite earlier ones, I'd add a third parameter `boolean overwrite` and `@returns if overwrite is false: true if the effect was registered, false if there already exists another effect for this transition; if overwrite is true: true if another registration was overwritten, false otherwise.`. A bit convoluted, so better find a nicer solution if needed...

Other opinions on this?

(PS: Same for the other register()s.)

Reply:
I'd prefer if other mods didn't overwrite screens as this is more intended for things like dimension mods to add for their own stuff. I will switch to the putIfAbsent thing, but if someone has an actual use for overwriting screens I will accommodate for that

Reply:
I think returning a boolean indicating whether the registration was successful would be better, but other than that I agree with this.